### ISSUE TITLE
[Feature] Conditional feed-forward operations

### ISSUE THREAD

**Issue by arthurfaria** [pasqal-io/qadence#453]
For my QGNN implementations, I need conditional operations since the quantum pooling layer relies on these types of gates [1].  As in CNNs, QCNNs consist of convolution layers that finds a new state and a pooling layer that reduces the size of the system. Below is an example:

<img width="300" alt="Picture1" src="https://github.com/pasqal-io/qadence/assets/12073321/78eabbcb-b335-4311-9bc7-96a6b6d75a9b">

Hence, for the quantum pooling layer, we need those **conditional feed-forward operations**,  defined as follows [2]:

- _Principle of deferred measurement:_ a controlled gate followed by a measurement of the controlled qubit results in the same as first performing the measurement and then applying a classically-controlled gate as in Fig. below

<img width="300" alt="Screenshot 2024-06-07 at 17 49 26" src="https://github.com/pasqal-io/qadence/assets/12073321/fe1549e1-8db8-41b7-ad01-62cabfdbc3bd">

In addition, these conditional operations are a necessary ingredient for quantum error correction, such as in surface code schemes. They check for errors using measurements in ancillas and depending on the measurement results, the system either performs error correction or does not.

-----------
ref. [1]: https://arxiv.org/pdf/2009.09423
ref. [2]: https://arxiv.org/pdf/2308.13065

**Comment by jpmoutinho** [comment]
Hey @arthurfaria, thanks for opening this! How urgently do you need it?

Doing a fully integrated implementation of conditional gates would likely require a bit of re-thinking of some core features of Qadence, like adding a classical register, a different measurement operation other than "sample", and implementing the control mechanism.

However, there should be sufficient tools in Qadence to replicate the circuit you want. For example, since the measured qubits are not reused, instead of measuring you can just do a normal control gate and then post-process the data at the end and post-select on the state of the ancilla qubits. That way each pooling layer will not "reduce" size of the system, but it will have the same effect. Happy to chat about this :)

**Comment by rajaiitp** [comment]
As far as the runtime, its gonna be identical as you would have to run thru all possible projections of the qubits being measured (subspace scaled down by 2**k, but additional 2**k simulations). If you are interested in specific projections alone, I presume a quicker hack would be is to use projector operator once all the controlled operations are performed and reuse the subspace output of interest for further simulations

**Comment by jpmoutinho** [comment]
Yes indeed @rajaiitp, and projectors are also available in Qadence (see [here](https://pasqal-io.github.io/qadence/v1.6.0/tutorials/advanced_tutorials/projectors/)).

**Comment by Roland-djee** [comment]
@jpmoutinho @arthurfaria @rajaiitp It would be good to start on this I guess.

**Comment by chMoussa** [comment]
Closing due to inactivity.
